File: tests/java/com/pdfjet/TextLineTest.java
/*
 * TextLineTest.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;

class TextLineTest {
    @Test
    void drawOnWritesTheTextAsHexAtTheFlippedY() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        TextLine line = new TextLine(TestSupport.helvetica(pdf), "Hello (x)").setLocation(10f, 20f);
        float[] xy = line.drawOn(page);
        String content = TestSupport.content(page);
        assertTrue(content.contains("10 772 Td\n"), content);
        assertTrue(content.contains("[<" + TestSupport.hex("Hello (x)") + ">] TJ\n"), content);
        assertEquals(44.664f, line.getWidth(), 0.001f);
        assertEquals(10f + line.getWidth(), xy[0], TestSupport.DELTA);
    }

    @Test
    void theUnderlineAndTheStrikeoutOfTaggedTextAreArtifacts() throws Exception {
        // The line is decoration: an element of its own, described as
        // "Underlined text: " and the text, is read after the text again.
        java.io.ByteArrayOutputStream bos = new java.io.ByteArrayOutputStream();
        PDF pdf = new PDF(bos, Compliance.PDF_UA_1);
        pdf.setTitle("Title");
        Page page = new Page(pdf, Letter.PORTRAIT);
        TextLine line = new TextLine(TestSupport.helvetica(pdf), "Hello");
        line.setUnderline(true);
        line.setStrikeout(true);
        line.setLocation(10f, 20f);
        line.drawOn(page);
        String content = TestSupport.content(page);
        assertEquals(2, content.split("/Artifact BMC\n", -1).length - 1, content);
        pdf.complete();
        String raw = TestSupport.latin1(bos.toByteArray());
        assertEquals(1, raw.split("/S /P\n", -1).length - 1, "the text line is more than one element");
        assertFalse(raw.contains("/Alt "), "the lines describe themselves");
    }

    @Test
    void emptyTextDrawsNothingAndReturnsTheLocation() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        TestSupport.assertXY(5f, 6f, new TextLine(TestSupport.helvetica(pdf), "").setLocation(5f, 6f).drawOn(page));
        assertEquals(0, page.getContent().length);
    }

    @Test
    void getLocationReturnsACopyOfTheLocation() throws Exception {
        TextLine line = new TextLine(TestSupport.helvetica(TestSupport.newPDF()), "x").setLocation(5f, 6f);
        line.getLocation()[1] = 60f;
        TestSupport.assertXY(5f, 6f, line.getLocation());
    }

    @Test
    void colorSettersConvertAndCopy() throws Exception {
        TextLine line = new TextLine(TestSupport.helvetica(TestSupport.newPDF()), "x");
        line.setTextColor(0xFF8000);
        TestSupport.assertRGB(1f, 128f / 255f, 0f, line.getTextColor());

        float[] rgb = {0.1f, 0.2f, 0.3f};
        line.setTextColor(rgb);
        rgb[0] = 0.9f;
        line.getTextColor()[1] = 0.9f;
        TestSupport.assertRGB(0.1f, 0.2f, 0.3f, line.getTextColor());
    }

    @Test
    void underlineAddsAStrokedLine() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Page plain = new Page(pdf, Letter.PORTRAIT);
        new TextLine(TestSupport.helvetica(pdf), "Hello").setLocation(10f, 20f).drawOn(plain);
        assertFalse(TestSupport.content(plain).contains("\nS\n"));

        Page underlined = new Page(pdf, Letter.PORTRAIT);
        new TextLine(TestSupport.helvetica(pdf), "Hello").setLocation(10f, 20f).setUnderline(true).drawOn(underlined);
        assertTrue(TestSupport.content(underlined).contains("\nS\n"));
    }

    @Test
    void setFontChangesTheFallbackFontUnlessAnotherWasSet() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Font helvetica = TestSupport.helvetica(pdf);
        Font courier = new Font(pdf, CoreFont.COURIER);
        TextLine line = new TextLine(helvetica, "x").setFont(courier);
        assertSame(courier, line.getFallbackFont());
        line.setFallbackFont(helvetica).setFont(new Font(pdf, CoreFont.TIMES_ROMAN));
        assertSame(helvetica, line.getFallbackFont());
    }
}

File: tests/dotnet/TextLineTest.cs
/*
 * TextLineTest.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using Xunit;

namespace PDFjet.NET {
public class TextLineTest {
    [Fact]
    public void DrawOnWritesTheTextAsHexAtTheFlippedY() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        TextLine line = new TextLine(TestSupport.Helvetica(pdf), "Hello (x)").SetLocation(10f, 20f);
        float[] xy = line.DrawOn(page);
        string content = TestSupport.Content(page);
        Assert.Contains("1 0 0 1 10 772 Tm\n", content);
        Assert.Contains("[<" + TestSupport.Hex("Hello (x)") + ">] TJ\n", content);
        TestSupport.AssertNear(44.664f, line.GetWidth(), 0.001f);
        TestSupport.AssertNear(10f + line.GetWidth(), xy[0], TestSupport.DELTA);
    }

    [Fact]
    public void EmptyTextDrawsNothingAndReturnsTheLocation() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        TestSupport.AssertXY(5f, 6f, new TextLine(TestSupport.Helvetica(pdf), "").SetLocation(5f, 6f).DrawOn(page));
        Assert.Empty(page.GetContent());
    }

    [Fact]
    public void ColorSettersConvertAndCopy() {
        TextLine line = new TextLine(TestSupport.Helvetica(TestSupport.NewPDF()), "x");
        line.SetTextColor(0xFF8000);
        TestSupport.AssertRGB(1f, 128f / 255f, 0f, line.GetTextColor());

        float[] rgb = {0.1f, 0.2f, 0.3f};
        line.SetTextColor(rgb);
        rgb[0] = 0.9f;
        line.GetTextColor()[1] = 0.9f;
        TestSupport.AssertRGB(0.1f, 0.2f, 0.3f, line.GetTextColor());
    }

    [Fact]
    public void UnderlineAddsAStrokedLine() {
        PDF pdf = TestSupport.NewPDF();
        Page plain = new Page(pdf, Letter.PORTRAIT);
        new TextLine(TestSupport.Helvetica(pdf), "Hello").SetLocation(10f, 20f).DrawOn(plain);
        Assert.DoesNotContain("\nS\n", TestSupport.Content(plain));

        Page underlined = new Page(pdf, Letter.PORTRAIT);
        new TextLine(TestSupport.Helvetica(pdf), "Hello").SetLocation(10f, 20f).SetUnderline(true).DrawOn(underlined);
        Assert.Contains("\nS\n", TestSupport.Content(underlined));
    }

    [Fact]
    public void SetFontChangesTheFallbackFontUnlessAnotherWasSet() {
        PDF pdf = TestSupport.NewPDF();
        Font helvetica = TestSupport.Helvetica(pdf);
        Font courier = new Font(pdf, CoreFont.COURIER);
        TextLine line = new TextLine(helvetica, "x").SetFont(courier);
        Assert.Same(courier, line.GetFallbackFont());
        line.SetFallbackFont(helvetica).SetFont(new Font(pdf, CoreFont.TIMES_ROMAN));
        Assert.Same(helvetica, line.GetFallbackFont());
    }
}
}
